server/src/main/java/org/elasticsearch/index/shard/IndexShardOperationPermits.java

```diff
@@ -128,21 +124,20 @@ private void delayOperations() {
         }
     }
 
-    private Releasable acquireAll(final long timeout, final TimeUnit timeUnit) throws InterruptedException {
+    private Releasable acquireAll() throws InterruptedException {
         if (Assertions.ENABLED) {
             // since delayed is not volatile, we have to synchronize even here for visibility
             synchronized (this) {
                 assert queuedBlockOperations > 0;
             }
         }
-        if (semaphore.tryAcquire(TOTAL_PERMITS, timeout, timeUnit)) {
-            return Releasables.releaseOnce(() -> {
-                assert semaphore.availablePermits() == 0;
-                semaphore.release(TOTAL_PERMITS);
-            });
-        } else {
-            throw new ElasticsearchTimeoutException("timeout while blocking operations after [" + new TimeValue(timeout, timeUnit) + "]");
+        while (semaphore.tryAcquire(TOTAL_PERMITS, 5, TimeUnit.MINUTES) == false) {
+            logger.warn("failed to acquire all indexing operation permits for {} after 5 minutes, retrying", shardId);
```

Comment:
Note that previously if there was more than one block pending then they would be serviced in FIFO order (the semaphore is fair) whereas with this retry loop we lose that ordering guarantee. I do not think this is important, we very rarely have any blocks pending let alone having more than one.